scripts/test_metric_reporter/circleci_json_parser.py

```diff
@@ -0,0 +1,93 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+"""Module for parsing test suite results from CircleCI metadata."""
+
+import json
+import os
+from typing import Any
+
+from pydantic import BaseModel, ValidationError
+
+from scripts.common.error import BaseError
+
+
+class CircleCIJob(BaseModel):
+    """Represents a CircleCI job with its metadata."""
+
+    dependencies: list[Any]
+    job_number: int
+    id: str
+    started_at: str
+    name: str
+    project_slug: str
+    status: str
+    type: str
+    stopped_at: str
+
+
+class CircleCITestMetadata(BaseModel):
+    """Represents test metadata for a CircleCI job."""
+
+    classname: str
+    name: str
+    result: str
+    message: str
+    run_time: float
+    source: str
+
+
+class CircleCIJobTestMetadata(BaseModel):
+    """Represents the test suite results for a CircleCI job."""
+
+    job: CircleCIJob
+    test_metadata: list[CircleCITestMetadata] | None = None
+
+
+class CircleCIJsonParserError(BaseError):
+    """Custom exception for errors raised by the CircleCI JSON parser."""
+
+    pass
+
+
+class CircleCIJsonParser:
+    """Parses CircleCI JSON test metadata files."""
+
+    @staticmethod
+    def parse(test_metadata_directory: str) -> list[CircleCIJobTestMetadata]:
+        """Parse CircleCI JSON test metadata from the specified directory.
+
+        Args:
+            test_metadata_directory (str): The path to the directory containing the test metadata
+                                           files.
+
+        Returns:
+            list[CircleCIJobTestMetadata]: A list of parsed CircleCIJobTestMetadata objects.
+
+        Raises:
+            CircleCIJsonParserError: If the directory does not exist, if there are errors reading
+                                     files, or if there are issues with parsing the JSON data.
+        """
+        if not os.path.isdir(test_metadata_directory):
+            raise CircleCIJsonParserError("The `test_metadata_directory` does not exist")
+
+        results: list[CircleCIJobTestMetadata] = []
+
+        try:
+            test_result_files: list[str] = sorted(os.listdir(test_metadata_directory))
+            for test_result_file in test_result_files:
+                test_result_file_path = os.path.join(test_metadata_directory, test_result_file)
+                with open(test_result_file_path) as file:
+                    data: dict[str, Any] = json.load(file)
```

Comment:
it will abort the for loop for all. Since we actually dictate the schema for CircleCI JSONs and write them, I figured it was safe to assume if one is bad they are all bad, but maybe that's not for the best. A ValidationError really shouldn't happen, it means we screwed up the circleci_scraper.

In the case of junit xmls i wouldn't make this assumption by contrast.